Optix coopvec matmul

szihs opened the pull request.

szihs commented:
/format

szihs commented:
/format

szihs commented:
@mkeshavaNV can you review this

mkeshavaNV commented on source/slang/hlsl.meta.slang:
```diff
@@ -26476,6 +26480,58 @@ uint32_t __getHLSLCoopVecMatrixLayout(CoopVecMatrixLayout layout)
 }
 
+[ForceInline]
+uint32_t __getOptixCoopVecComponentType(CoopVecComponentType componentType)
```
where does this get used?

jkwak-work commented on source/slang/hlsl.meta.slang:
```diff
@@ -26674,15 +26731,14 @@ CoopVec<T, M> coopVecMatMulPacked(
     constexpr uint matrixStride
 )
 {
-    static_assert(__isPackedInputInterpretation(inputInterpretation) || k == PackedK
-                 , "for non-packed inputInterpretation values k must be equal to the input vector length");
-    static_assert(!__isPackedInputInterpretation(inputInterpretation)
-                || k <= __inputInterpretationPackingFactor(inputInterpretation)*PackedK
-                 , "for packed inputInterpretation values k must be less than or equal to the input vector length times the packing factor");
-
```
I am not sure why these static_asserts don't apply to optiX.
When the given value is not-packed, the count should match, isn't it?
And if it is packed, the given packed-value should be smaller than the actual count.

szihs replied:
If not - I suggest to take this as separate issue (unrelated to optiX) -  https://github.com/shader-slang/slang/pull/8489#issuecomment-3355598699

jkwak-work replied:
If the static_assert is not working, we should fix it.
When there is an assertion failure, you should fix the problem not removing the assertion.

jkwak-work replied:
The explanation from LLM makes no sense because the branching behavior is same to all platforms.
The example LLM shows is following,
```
block blockMerge(param %isPacked : Bool): ← Phi node/block parameter
static_assert(%isPacked || ..., "...")
```
In this example, `%isPacked` is supposed to be constant folded in a function `applySparseConditionalConstantPropagation` as a part of function call to `simplifyIR()`.

Consider the following examples and please investigate more.

The following command produced CUDA fine:
```
build/Release/bin/slangc.exe -stage compute -entry computeMain -target cuda tests/cooperative-vector/matrix-mul-packed.slang

... no errors ...
```

But the following command prints errors:
```
build/Release/bin/slangc.exe -stage compute -entry computeMain -target cuda tests/cooperative-vector/matrix-mul-packed.slang -capability optix_coopvec
hlsl.meta.slang(26727): error 41402: condition for static assertion cannot be evaluated at compile time.
hlsl.meta.slang(26730): error 41402: condition for static assertion cannot be evaluated at compile time.
```
Why is it working for CUDA and not working when `-capability optix_coopvec` is used?
It think something is not right when the capability `optix_coopvec` is used.

jkwak-work reviewed (commented):
Do you have a github issue number for this?
Can you mention that in the PR description?

jkwak-work reviewed (commented):
The new code change keeps static_assert with the following workaround.
```
  coopVecMatMul wrapper has:
  - static_assert(condition || K == K, "...")  [NO GenericAsm]

  __coopVecMatMulPacked_impl has:
  - __intrinsic_asm [NO static assertions]
```

However, it will break all of the documentation work done for the functions. In other words, the comments on `__coopVecMatMulPacked_impl` should be for `coopVecMatMul`.
And I think `__coopVecMatMulPacked_impl` should have `internal` visibility keyword.

We should also have a new github issue to track the limitation about this and put the issue number as a comment somewhere.

szihs commented:
> The new code change keeps static_assert with the following workaround.
> 
> ```
>   coopVecMatMul wrapper has:
>   - static_assert(condition || K == K, "...")  [NO GenericAsm]
> 
>   __coopVecMatMulPacked_impl has:
>   - __intrinsic_asm [NO static assertions]
> ```
> 
> However, it will break all of the documentation work done for the functions. In other words, the comments on `__coopVecMatMulPacked_impl` should be for `coopVecMatMul`. And I think `__coopVecMatMulPacked_impl` should have `internal` visibility keyword.
> 
> We should also have a new github issue to track the limitation about this and put the issue number as a comment somewhere.



Agreed, Updated docs and added GitHub issue in the PR

szihs commented on source/slang/hlsl.meta.slang:
```diff
@@ -26861,6 +26849,61 @@ CoopVec<T, M> coopVecMatMulPacked(
+/// values to use from the packed input.
+// TODO: Can we ForceInline for just hlsl? the other platforms don't really
+// need it
+// Public wrapper with assertions (can be inlined since it has no GenericAsm)
```
Updated

szihs commented on source/slang/hlsl.meta.slang:
```diff
@@ -27658,6 +27774,7 @@ CoopVec<T, M> coopVecMatMulAddPacked<T : __BuiltinArithmeticType, let M : int, l
     constexpr uint matrixStride
 )
 {
+    // Common static assertions for all targets (moved before __target_switch)
```
Removed

szihs commented on source/slang/hlsl.meta.slang:
```diff
@@ -26700,27 +26704,14 @@ static const struct {
-// need it
+// Internal implementation without assertions (contains GenericAsm which prevents inlining)
+// TODO - https://github.com/shader-slang/slang/issues/8620
 [ForceInline]
```
Updated.

szihs commented:
> I left a few comments.

Updated commit and replied on the query.

jkwak-work commented on source/slang/slang-emit-cuda.cpp:
```diff
@@ -214,6 +214,24 @@ SlangResult CUDASourceEmitter::calcTypeName(IRType* type, CodeGenTarget target,
+            SLANG_DIAGNOSE_UNEXPECTED(
+                getSink(),
+                SourceLoc(),
+                "cooperative vector types are only supported for OptiX targets");
```
This statement itself isn't correct, because cooperative vector type is supported also for HLSL and SPIRV.

szihs replied:
updated
